sdks/java/core/src/test/java/org/apache/beam/sdk/transforms/reflect/DoFnSignaturesTest.java

```diff
@@ -401,6 +504,95 @@ public void foo(ProcessContext context) {}
   }
 
   @Test
+  public void testUsageOfStateDeclaredInSuperclass() throws Exception {
+    thrown.expect(IllegalArgumentException.class);
+    thrown.expectMessage("process");
+    thrown.expectMessage("declared in a different class");
+    thrown.expectMessage(DoFnDeclaringState.STATE_ID);
+    DoFnSignature sig =
+        DoFnSignatures.INSTANCE.getSignature(
+            new DoFnDeclaringState() {
+              @ProcessElement
+              public void process(
+                  ProcessContext context,
+                  @StateId(DoFnDeclaringState.STATE_ID) ValueState<Integer> state) {}
+            }.getClass());
+  }
+
+  @Test
+  public void testDeclOfStateUsedInSuperclass() throws Exception {
+    thrown.expect(IllegalArgumentException.class);
+    thrown.expectMessage("process");
+    thrown.expectMessage("declared in a different class");
+    thrown.expectMessage(DoFnUsingState.STATE_ID);
+    DoFnSignature sig =
+        DoFnSignatures.INSTANCE.getSignature(
+            new DoFnUsingState() {
```

Comment:
I think we can create named classes within the function, right? That would make these tests a bit easier to read, and I think the base classes only exist for one test.